### PR TITLE
network: serialize and deserialize current configuration

### LINKED ISSUE #26602: NFS root on IPv6 fails to boot
### systemd version the issue has been seen with

252

### Used distribution

Fedora 37

### Linux kernel version used

6.0.8-200.fc36.x86_64

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

A clean boot.

### Unexpected behaviour you saw

During boot, after switch-root, the system hangs. It some times eventually completes to a login prompt, but with many failures caused by stale file handles.

### Steps to reproduce the problem

Configure an initrd for nfs root (dracut customization file included).
Configure the network for IPv6. A sample systemd.network file included. It doesn't seem to matter whether it is configured by route advertisements or DHCPv6 (as per this example).

Boot with a command line like `root=nfs4:<server>:<exported fs>`

[config.zip](https://github.com/systemd/systemd/files/10839159/config.zip)

### Additional program output to the terminal or log subsystem illustrating the issue

https://gist.github.com/iandall/746bd4546f7629d15f39483825204772

### LINKED ISSUE #32569: SLAAC addrs perceived as foreign and removed during systemd-network start
### systemd version the issue has been seen with

255

### Used distribution

CentOS9

### Linux kernel version used

5.19

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

In a server environment, I want to be able to restart networkd and (assuming the network configuration is already set up), have networkd not change anything which would break IP connectivity.  Particularly, if an IP address is configured on an interface, and that's the expected IP, it should be left intact rather than deleting and re-adding it.  For static IPs I've used `KeepConfiguration = static` to achieve this, and I can restart networkd with no impact.

I need this to work for SLAAC addrs as well; upon restart of networkd, if a valid SLAAC addr is already configured on the link, it should be left intact and not deleted and re-added.

### Unexpected behaviour you saw

During restart of systemd-networkd, particularly during start, I can see that already-configured, not-expired SLAAC addrs are deleted off of the interface (observable via `ip monitor`).  With debug enabled the logs indicate the addr is considered foreign:

> eth6: Removing foreign address (configured,marked): REDACTED/64 (valid for 4w 1d 23h 59min 58s, preferred for 6d 23h 59min 58s), flags: manage-temporary-address,no-prefixroute, scope: global
...
> eth6: Forgetting foreign address (n/a): REDACTED/64 (valid for 4w 1d 23h 59min 58s, preferred for 6d 23h 59min 58s), flags: manage-temporary-address,no-prefixroute, scope: global

### Steps to reproduce the problem

Configure a SLAAC interface:

```
[Network]
IPv6AcceptRA = true
DHCP = no
KeepConfiguration = static

[IPv6AcceptRA]
UseAutonomousPrefix = true
UseOnLinkPrefix = true
DHCPv6Client = false
```

After the RA results in the interface getting a SLAAC addr, `systemctl restart systemd-networkd` and either via `ip monitor` or the debug logs, observe that networkd (temporarily) removes the addr.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Replaces #34963.

Fixes #26602.
Fixes #32569.

### CODE REVIEW

**Comment by bluca** [comment]
> This also drops meaningless dhcp-on-stop setting, as even if it is
specified, addresses or routes will be dropped on next invocation
anyway. Hence, the value is completely meaningless.

I am using this option - the manpage says:

> When set to "dhcp-on-stop", systemd-networkd will not drop addresses and routes on stopping the daemon.
> When "dhcp", the addresses and routes provided by a DHCP server will never be dropped even if the DHCP lease expires.

which seems to be different enough? IE, I want configuration to remain while networkd is restarted, but I want normal DHCP expirations rules to be followed.

**Comment by yuwata** [comment]
> > This also drops meaningless dhcp-on-stop setting, as even if it is
> > specified, addresses or routes will be dropped on next invocation
> > anyway. Hence, the value is completely meaningless.
> 
> I am using this option - the manpage says:
> 
> > When set to "dhcp-on-stop", systemd-networkd will not drop addresses and routes on stopping the daemon.
> > When "dhcp", the addresses and routes provided by a DHCP server will never be dropped even if the DHCP lease expires.
> 
> which seems to be different enough? IE, I want configuration to remain while networkd is restarted, but I want normal DHCP expirations rules to be followed.

Yeah, As you can see the CI result, the change breaks existing setups. I changed my mind, and I will keep the setting in the next push. Thank you for the comment.

**Comment by yuwata** [comment]
@anitazha @DaanDeMeyer With this PR, networkd keeps IPv6 connectivity (at least addresses and routes). If you'd like please test this, though still WIP and I may be changed details later. But, the basic behavior should be fine, at least brief test on my local machine.

File: src/network/networkd.c [hunk 1]
```diff
@@ -100,6 +101,10 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 return r;
 
+        r = manager_deserialize(m);
+        if (r < 0)
+                log_warning_errno(r, "Failed to deserialize the prvious invocation, ignoring: %m");
```
Comment (bluca) [inline src/network/networkd.c]: "previous"

File: src/network/networkd-ipv4ll.c [hunk 2]
```diff
@@ -193,6 +195,27 @@ static int ipv4ll_set_address(Link *link) {
...
+                if (a->family != AF_INET)
+                        continue;
+                if (!in4_addr_is_link_local_dynamic(&a->in_addr.in))
+                        continue;
+
+                r = sd_ipv4ll_set_address(link->ipv4ll, &a->in_addr.in);
+                if (r < 0)
+                        return r;
+
+                /* Make the address not removed by link_drop_unmanaged_addresses(). */
```
Comment (bluca) [inline src/network/networkd-ipv4ll.c]: "Make sure the address is not removed..."

File: man/systemd.network.xml [hunk 3]
```diff
@@ -1286,21 +1286,21 @@ DuplicateAddressDetection=none</programlisting></para>
...
+          dynamically configured addresses and routes will never be dropped, and the lifetime of DHCPv4
+          leases will be ignored. This is contrary to the DHCP specification, but may be the best choice if,
+          e.g., the root filesystem relies on this connection. The setting <literal>dynamic</literal> implies
+          <literal>dynamic-on-stop</literal>, and <literal>yes</literal> implies <literal>dynamic</literal>
+          and <literal>static</literal>. Defaults to <literal>dynamic-on-stop</literal> when
+          <command>systemd-networkd</command> is running in initrd, <literal>yes</literal> when the root
+          filesystem is a network filesystem, and <literal>no</literal> otherwise.</para>
 
-        <xi:include href="version-info.xml" xpointer="v243"/>
+          <xi:include href="version-info.xml" xpointer="v243"/>
```
Comment (bluca) [inline man/systemd.network.xml]: this should be v257 given the values have changed

**Comment by yuwata** [comment]
@bluca Thank you for the review. Addressed all requests. Upgrading the green label.

**Pull request merged**